contract_discount/models/contract.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+from openerp import models, fields, api
+from openerp.addons.decimal_precision import decimal_precision as dp
+
+
+class AccountAnalyticInvoiceLine(models.Model):
+    _inherit = "account.analytic.invoice.line"
+
+    @api.one
+    @api.depends('price_unit', 'discount')
+    def _amount_line(self):
```

Comment:
You shouldn't overwrite totally the compute method because you're altering the inheritance chain (imagine a second module that add a surcharge plus your discount..). Instead, you should keep old style function method, call super, and apply the discount over each line.